### PR TITLE
Add metadata for functions and aggregates

### PR DESCRIPTION
**Pull request by beltran**

### CODE REVIEW

File: metadata.go [hunk 1]
```diff
@@ -216,13 +253,25 @@ func compileMetadata(
...
 	}
+	keyspace.Functions = make(map[string]*FunctionMetadata, len(functions))
+	for i := range functions {
+		keyspace.Functions[functions[i].Name] = &functions[i]
+	}
+	keyspace.Aggregates = make(map[string]*AggregateMetadata, len(aggregates))
+	for i, aggregate := range aggregates {
+		aggregate.FinalFunc = *keyspace.Functions[aggregate.finalFunc]
+		aggregate.StateFunc = *keyspace.Functions[aggregate.stateFunc]
+		keyspace.Aggregates[aggregates[i].Name] = &aggregate
```
Comment (alourie) [inline metadata.go]: isn't `aggregates[i] == aggregate` ? Also, you could use the same `for i, function := range functions` above....

File: cassandra_test.go [hunk 2]
```diff
@@ -2234,6 +2234,126 @@ func TestGetColumnMetadata(t *testing.T) {
...
+				NativeType{typ: TypeBigInt},
+			},
+		},
+		stateFunc: "avgstate",
+		finalFunc: "avgfinal",
+	}
+
+	// In this case cassandra is returning a blob
+	if  flagCassVersion.Before(3, 0, 0) {
+		expectedAggregrate.InitCond = string([]byte{0, 0 ,0 , 4, 0, 0, 0, 0, 0, 0, 0,  8, 0, 0, 0, 0, 0, 0, 0, 0})
```
Comment (alourie) [inline cassandra_test.go]: few spaces are in the wrong spots

**Comment by beltran** [comment]
Thank you for the review @alourie .

File: metadata.go [hunk 3]
```diff
@@ -216,13 +253,25 @@ func compileMetadata(
...
+	aggregates []AggregateMetadata,
 ) {
 	keyspace.Tables = make(map[string]*TableMetadata)
 	for i := range tables {
 		tables[i].Columns = make(map[string]*ColumnMetadata)
 
 		keyspace.Tables[tables[i].Name] = &tables[i]
 	}
+	keyspace.Functions = make(map[string]*FunctionMetadata, len(functions))
+	for i := range functions {
```
Comment (alourie) [inline metadata.go]: Pretty please :-) ?

Just let's make it consistent with 
```golang
for _, function := range functions {
        keyspace.Functions[function.Name] = &function
}
```

Comment (beltran) [reply metadata.go]: For some reason the tests fail with this change.

Comment (alourie) [reply metadata.go]: yep, that's the reason.

File: metadata.go [hunk 4]
```diff
@@ -793,6 +842,129 @@ func getColumnMetadata(session *Session, keyspaceName string) ([]ColumnMetadata,
...
+			&function.Language,
+			&returnType,
+		)
+		function.ReturnType = getTypeInfo(returnType)
+		function.ArgumentTypes = make([]TypeInfo, len(argumentTypes))
+		for i, argumentType := range argumentTypes {
+			function.ArgumentTypes[i] = getTypeInfo(argumentType)
+		}
+
+		if err != nil {
```
Comment (Zariel) [inline metadata.go]: move the error check to where the error is returned, line 888

File: metadata.go [hunk 5]
```diff
@@ -793,6 +842,129 @@ func getColumnMetadata(session *Session, keyspaceName string) ([]ColumnMetadata,
...
+		)
+
+		aggregate.ReturnType = getTypeInfo(returnType)
+		aggregate.StateType = getTypeInfo(stateType)
+		aggregate.ArgumentTypes = make([]TypeInfo, len(argumentTypes))
+		for i, argumentType := range argumentTypes {
+			aggregate.ArgumentTypes[i] = getTypeInfo(argumentType)
+		}
+
+		if err != nil {
```
Comment (Zariel) [inline metadata.go]: again move error check to where the error is returned, line 947

**Comment by alourie** [comment]
We just see random errors now....will need to investigate :-)

**Comment by beltran** [comment]
hmm, the failure seems unrelated.

**Comment by alourie** [comment]
@beltran would you mind rebasing with master and pushing again? Just to give the CI a nudge to try again. I don't mind merging it though.

**Comment by beltran** [comment]
👍 sure

**Pull request merged**